### ISSUE TITLE
[FR] Add ability to connect to a single host using different keys per repo/remote

### ISSUE THREAD

**Issue by tp-woven** [martinvonz/jj#1455]
At work, we are migrating to use GitHub's [Enterprise Managed Users](https://docs.github.com/en/enterprise-cloud@latest/admin/identity-and-access-management/using-enterprise-managed-users-for-iam/about-enterprise-managed-users) (EMU), and our enterprise policy prevents enterprise users from interacting with non-enterprise repos (no PRs, no commenting on issues, etc.). So in order to contribute to OSS, we have to maintain two separate users - an enterprise one and a public one. This causes a problem when using SSH keys to authenticate - GitHub does not allow the same key to be used for multiple users, so each user has to have its own SSH key.

This then brings up the question of how to choose which key to use when connecting to github.com. With Git, it respects `~/.ssh/config`, so the following setup works (just have to remember to use the appropriate host in remotes):

```
Host github
  HostName github.com
  User tp-woven
  IdentityFile ~/.ssh/id_ed25519
  AddKeysToAgent yes
  IdentitiesOnly yes

Host github-sg
  HostName github.com
  User <enterprise_user>
  IdentityFile ~/.ssh/id_ed25519_emu
  AddKeysToAgent yes
  IdentitiesOnly yes
```

However, trying to use this setup with `jj` results in an error:

```
Error: failed to resolve address for github-sg: nodename nor servname provided, or not known; class=Net (12)
```

@yuja suggested on [Discord](https://discord.com/channels/968932220549103686/969829516539228222/1090573050661048371) that I may be able to use `ssh-add -h`, but this does not seem to be supported on macOS.

**Comment by xelxebar** [comment]
The issue is that authentication just calls out to libssl2 (because that's what libgit2 uses), which does no `.ssh/config` parsing at all. In general, to get the desired behavior you describe, 2 main pieces are needed:

1. Embed user (and port) overrides inside your remote URL: e.g. `https://tp-woven@github.com/...`,
2. Copy your keys into `ssh-agent`.

It sounds like you might just be missing 1, perhaps?

**Comment by tp-woven** [comment]
@xelxebar AFAIK, the user for SSH remotes always needs to be `git` (at least on GH, changing from `git@github.com` to `tp-woven@github.com` appears to break the remote)

**Comment by xelxebar** [comment]
Hrm... I played around with the `User` setting in `.ssh/config`. It looks like the `git@` in the URL overrides it, so my suspicion is that the proper key just isn't getting selected. What happens if you temporarily remove all other keys and just add your github ones to `ssh-agent`?

**Comment by tp-woven** [comment]
Sorry, I'm not sure I follow, I think that's what I am already doing?

I'm also not entirely sure how GH/Git "server" handles the identity, but when I set up my `.ssh/config` as I mentioned above, I can do this:

```
❯ ssh -T git@github
Hi tp-woven! You've successfully authenticated, but GitHub does not provide shell access.
❯ ssh -T git@github-sg
Hi <enterprise_user>! You've successfully authenticated, but GitHub does not provide shell access.
```

Note that in both cases, I use `git@`, but the `User` definition from `.ssh/config` apparently works.

**Comment by tp-woven** [comment]
Oh, I see what you mean, yeah - removing `User` does still work. Similarly, only having a single key loaded in my agent works too. Are you suggesting that I remove and add the keys every time I push/pull? I would like to be able to work on both public and "enterprise" repos "at the same time", so that might become a bit tedious. Or maybe you mean I should run multiple `ssh-agent`s and choose between them using env variables? I didn't even realize that was an option 😅, but it does look like there is some documentation on how to do that... (Not sure if it's something I can actually do on my corp laptop, though, I'll have to look into that)

> Apologies for the rough and terse messages. I'm writing these in the small gaps of time between work issues.

No worries, I'm grateful for the help :)

**Comment by xelxebar** [comment]
> Similarly, only having a single key loaded in my agent works too.

Okay, good. So the fundamental connection parameters are all definitely in place.

> Are you suggesting that I ...

Honestly, I can't think of a really good suggestion. The two you lay out are (annoying) options. Running separate `ssh-agent` instances is simple enough—setting `SSH_AUTH_SOCK` lets you choose which one clients should talk to. With appropriate shell aliases, this might be non-horrible.

**Comment by tp-woven** [comment]
> setting SSH_AUTH_SOCK lets you choose which one clients should talk to. With appropriate shell aliases, this might be non-horrible.

Yeah, maybe with `direnv` or something like that it won't be too bad... But it remains to be seen whether I can actually set it up here. :/

Anyway, thanks for the help and suggestions! :)

**Comment by yuja** [comment]
It might be better to use https and git credential helper. I don't know details, but appears that it can use the system keyring on macOS?

https://docs.github.com/en/get-started/getting-started-with-git/updating-credentials-from-the-macos-keychain

**Comment by tp-woven** [comment]
Ah, interesting, let me try that, thanks!

**Comment by steveklabnik** [comment]
I *think* because we now shell out to `git`, this should Just Work?
